compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2003,20 +2003,29 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
         // local adaptation makes sure every adapted tree conforms to its pt
         // so will take the code path that decides on inlining
         return tpd.Block(adapt(tree, WildcardType) :: Nil, Literal(Constant(())))
-      // convert function literal to SAM closure
       tree match {
-        case Closure(Nil, id @ Ident(nme.ANON_FUN), _)
-        if defn.isFunctionType(wtp) && !defn.isFunctionType(pt) =>
-          pt match {
-            case SAMType(meth)
-            if wtp <:< meth.info.toFunctionType() =>
-              // was ... && isFullyDefined(pt, ForceDegree.noBottom)
-              // but this prevents case blocks from implementing polymorphic partial functions,
-              // since we do not know the result parameter a priori. Have to wait until the
-              // body is typechecked.
-              return cpy.Closure(tree)(Nil, id, TypeTree(pt)).withType(pt)
-            case _ =>
-          }
+        case Closure(Nil, id @ Ident(nme.ANON_FUN), _) =>
+          val adaptToPrototype =
+            if (defn.isFunctionType(pt)) {
+              // convert implicit function to function
```

Comment:
Not sure about the change in this logic. Questions:

 - Do we want to convert implicit functions to functions, or to other SAM types? Does this case even arise
    as an implicit function is always immediately applied?
 - What would happen if we kept the previous code here?